apps/emqx/src/emqx_persistent_message_ds_replayer.erl

```diff
@@ -75,59 +94,81 @@ next_packet_id(Inflight0 = #inflight{next_seqno = LastSeqno}) ->
     {seqno_to_packet_id(LastSeqno), Inflight}.
 
 -spec n_inflight(inflight()) -> non_neg_integer().
-n_inflight(#inflight{next_seqno = NextSeqno, acked_until = AckedUntil}) ->
-    range_size(AckedUntil, NextSeqno).
+n_inflight(#inflight{offset_ranges = Ranges}) ->
+    %% TODO
```

Comment:
Just a vague idea: can we have two ephemeral (non-persistent) counters for this? One is for QoS1 messages, and another is for QoS2.
- Once the process reconnects, we reconstruct the value of this counters by scanning the ranges.
- Otherwise, we just update it with each message. PUBACK for QoS1 message decrements QoS1 counter; PUBCOMP for QoS2 message decrements QoS2 counter.
- n_inflight = counter_qos1 + counter_qos2.

I see the problem that it requires storing some properties (like QoS) for each outgoing message. But if we hold onto the batch for longer (we probable should, since it's required to process hooks for PUBACK), we could get this information from the batch itself.